src/server/project.ts

```diff
@@ -1572,69 +1594,59 @@ namespace ts.server {
-
-            return result;
+            return this.projectService.getPackageJsonsVisibleToFile(fileName, rootDir);
         }
 
         /*@internal*/
-        onAddPackageJson(path: Path) {
-            this.packageJsonCache.addOrUpdate(path);
-            this.watchPackageJsonFile(path);
+        getPackageJsonsForAutoImport(rootDir?: string): readonly PackageJsonInfo[] {
+            const packageJsons = this.getPackageJsonsVisibleToFile(combinePaths(this.currentDirectory, inferredTypesContainingFile), rootDir);
+            this.packageJsonsForAutoImport = arrayToSet(packageJsons.map(p => p.fileName));
+            return packageJsons;
         }
 
         /*@internal*/
         getImportSuggestionsCache() {
             return this.importSuggestionsCache;
         }
 
-        private watchPackageJsonFile(path: Path) {
-            const watchers = this.packageJsonFilesMap || (this.packageJsonFilesMap = createMap());
-            if (!watchers.has(path)) {
-                watchers.set(path, this.projectService.watchFactory.watchFile(
-                    this.projectService.host,
-                    path,
-                    (fileName, eventKind) => {
-                        const path = this.toPath(fileName);
-                        switch (eventKind) {
-                            case FileWatcherEventKind.Created:
-                                return Debug.fail();
-                            case FileWatcherEventKind.Changed:
-                                this.packageJsonCache.addOrUpdate(path);
-                                break;
-                            case FileWatcherEventKind.Deleted:
-                                this.packageJsonCache.delete(path);
-                                watchers.get(path)!.close();
-                                watchers.delete(path);
-                        }
-                    },
-                    PollingInterval.Low,
-                    this.projectService.getWatchOptions(this),
-                    WatchType.PackageJsonFile,
-                ));
+        /*@internal*/
+        includePackageJsonAutoImports(): PackageJsonAutoImportPreference {
+            if (this.projectService.includePackageJsonAutoImports() === PackageJsonAutoImportPreference.None ||
+                !this.languageServiceEnabled ||
+                isInsideNodeModules(this.currentDirectory) ||
+                !this.isDefaultProjectForOpenFiles()) {
+                return PackageJsonAutoImportPreference.None;
             }
+            return this.projectService.includePackageJsonAutoImports();
+        }
+
+        /*@internal*/
+        getPackageJsonAutoImportProvider(): Program | undefined {
+            if (this.autoImportProviderHost) {
+                updateProjectIfDirty(this.autoImportProviderHost);
+                if (!this.autoImportProviderHost.hasRoots()) {
+                    this.autoImportProviderHost.close();
+                    this.autoImportProviderHost = undefined;

```

Comment:
I think instead of setting this as undefined here you want to set it to false or something so that next time some one asks for project, you dont have to go through dependencies, resolving type references etc unless there is some change. So you want this to autoImportProviderHost to be undefined if not calculated, false if calculated but shouldnt use and Actual project if it should